File: examples/rust/import_resolution_test.rs
//! Test file specifically for verifying import resolution
//! This file tests that imports are correctly resolved to their definitions
//!
//! EXPECTED BEHAVIOR:
//! 1. Standard library imports should resolve to std:: paths
//! 2. Aliased imports should track both original and alias names
//! 3. Internal module imports should resolve relative to this file
//! 4. Nested module imports should resolve through the hierarchy
//! 5. Super imports should resolve to parent modules
//! 6. Crate-level imports should resolve absolutely

// === STANDARD LIBRARY IMPORTS ===
// These should resolve to their std:: paths
use std::collections::HashMap;
use std::sync::Arc;

// === ALIASED IMPORTS ===
// These should track both the original path and the alias
use std::collections::HashSet as Set;
use std::sync::Mutex as Lock;

// === INTERNAL MODULE STRUCTURE ===
// Module for testing internal imports
mod helpers {
    pub fn helper_function() -> String {
        "Helper".to_string()
    }

    pub struct HelperStruct {
        pub value: i32,
    }

    impl HelperStruct {
        pub fn new(value: i32) -> Self {
            HelperStruct { value }
        }
    }

    // Nested module for testing deeper imports
    pub mod nested {
        pub fn nested_function() -> String {
            "Nested".to_string()
        }

        // Test super import (importing from parent module)
        use super::HelperStruct;

        pub fn use_parent_struct() -> HelperStruct {
            HelperStruct::new(100)
        }
    }
}

// === RELATIVE MODULE IMPORTS ===
// Import from internal module - should resolve to helpers:: within this file's context
use helpers::{helper_function, HelperStruct};
use helpers::nested::nested_function;

// === CONFLICTING NAMES TEST ===
// Another module that also has a helper_function (to test resolution picks the right one)
mod other_helpers {
    pub fn helper_function() -> i32 {
        42
    }

    // This import should NOT be confused with helpers::helper_function
    pub fn call_own_helper() -> i32 {
        helper_function()  // Calls other_helpers::helper_function
    }
}

// === CRATE-LEVEL IMPORTS (if this were a multi-file crate) ===
// These demonstrate how imports would work in a real crate
// use crate::helpers::HelperStruct;  // Would work if we needed explicit crate import

// === SELF IMPORTS ===
// Import from self (current module)
use self::other_helpers::helper_function as other_helper;

// === MAIN FUNCTION WITH IMPORT USAGE ===
fn main() {
    println!("=== Testing Import Resolution ===\n");

    // TEST 1: Standard library imports
    println!("1. Standard library imports:");
    let mut map: HashMap<String, i32> = HashMap::new();
    map.insert("key".to_string(), 42);
    println!("   HashMap created and used ✓");

    let data = Arc::new(vec![1, 2, 3]);
    println!("   Arc created and used ✓");

    // TEST 2: Aliased imports
    println!("\n2. Aliased imports:");
    let mut set: Set<i32> = Set::new();  // Using alias 'Set' for HashSet
    set.insert(1);
    println!("   HashSet used as 'Set' ✓");

    let lock = Lock::new(5);  // Using alias 'Lock' for Mutex
    println!("   Mutex used as 'Lock' ✓");

    // TEST 3: Relative module imports
    println!("\n3. Relative module imports:");
    // This should resolve to helpers::helper_function, NOT other_helpers::helper_function
    let result = helper_function();
    println!("   helper_function() returns: '{}' (expected: 'Helper') ✓", result);

    let helper = HelperStruct::new(10);
    println!("   HelperStruct::new() works with value: {} ✓", helper.value);

    // TEST 4: Nested module import
    println!("\n4. Nested module imports:");
    let nested_result = nested_function();
    println!("   nested_function() returns: '{}' (expected: 'Nested') ✓", nested_result);

    // TEST 5: Self import with alias
    println!("\n5. Self imports with alias:");
    let other_result = other_helper();
    println!("   other_helper() returns: {} (expected: 42) ✓", other_result);

    println!("\n=== All import tests completed ===");
}

// === TEST MODULE WITH SUPER IMPORTS ===
#[cfg(test)]
mod tests {
    // Import everything from parent module
    use super::*;

    #[test]
    fn test_standard_imports() {
        // HashMap should be available through super::*
        let _map: HashMap<String, i32> = HashMap::new();
        // Arc should be available
        let _arc: Arc<Vec<i32>> = Arc::new(vec![]);
    }

    #[test]
    fn test_aliased_imports() {
        // Aliases should work through super::*
        let _set: Set<i32> = Set::new();
        let _lock: Lock<i32> = Lock::new(0);
    }

    #[test]
    fn test_relative_imports() {
        // This should resolve to helpers::helper_function due to the import
        let result = helper_function();
        assert_eq!(result, "Helper", "Should call helpers::helper_function");

        // This should resolve to helpers::HelperStruct::new
        let helper = HelperStruct::new(20);
        assert_eq!(helper.value, 20);

        // Nested function should work
        let nested = nested_function();
        assert_eq!(nested, "Nested");
    }

    #[test]
    fn test_self_imports() {
        // Test the aliased import
        let result = other_helper();
        assert_eq!(result, 42, "Should call other_helpers::helper_function via alias");
    }

    // Nested test module to test deeper super imports
    mod nested_tests {
        use super::super::*;  // Import from grandparent

        #[test]
        fn test_grandparent_imports() {
            // Should still have access to helpers through grandparent
            let result = helper_function();
            assert_eq!(result, "Helper");
        }
    }
}

// === RESOLUTION EXPECTATIONS SUMMARY ===
//
// When indexed and analyzed, we expect:
//
// 1. STANDARD LIBRARY:
//    - HashMap resolves to std::collections::HashMap
//    - Arc resolves to std::sync::Arc
//
// 2. ALIASES:
//    - Set resolves to std::collections::HashSet (with alias tracked)
//    - Lock resolves to std::sync::Mutex (with alias tracked)
//
// 3. RELATIVE MODULES:
//    - helpers::helper_function resolves to
//      crate::examples::rust::import_resolution_test::helpers::helper_function
//    - helpers::HelperStruct resolves to
//      crate::examples::rust::import_resolution_test::helpers::HelperStruct
//
// 4. NESTED MODULES:
//    - helpers::nested::nested_function resolves to
//      crate::examples::rust::import_resolution_test::helpers::nested::nested_function
//
// 5. SUPER IMPORTS:
//    - super::HelperStruct (in nested module) resolves to parent's HelperStruct
//    - super::* (in tests) imports everything from parent module
//
// 6. SELF IMPORTS:
//    - self::other_helpers::helper_function resolves within current module
//
// 7. DISAMBIGUATION:
//    - helper_function() in main calls helpers::helper_function (via import)
//    - helper_function() in other_helpers calls other_helpers::helper_function (no import)

File: examples/go/import_resolution_test.go
//! Test file specifically for verifying import resolution
//! This file tests that imports are correctly resolved to their definitions
//!
//! EXPECTED BEHAVIOR:
//! 1. Standard library imports should resolve to their standard packages
//! 2. Aliased imports should track both original and alias names
//! 3. Internal package imports should resolve relative to this module
//! 4. Nested package imports should resolve through the hierarchy
//! 5. Dot imports should import into current namespace
//! 6. Package-level imports should resolve absolutely

package main

// === STANDARD LIBRARY IMPORTS ===
// These should resolve to their standard library paths
import (
	"fmt"
	"sync"
	"time"
)

// === ALIASED IMPORTS ===
// These should track both the original path and the alias
import (
	"context"
	set "container/list" // Using list as set-like structure
	lock "sync"          // Aliasing sync as lock
)

// === INTERNAL PACKAGE STRUCTURE ===
// Helpers package equivalent (in Go, this would be a separate package)
// For this example, we'll simulate it with types and functions in the same package

// Helper functions (equivalent to helpers module)
func helperFunction() string {
	return "Helper"
}

type HelperStruct struct {
	Value int
}

func NewHelperStruct(value int) *HelperStruct {
	return &HelperStruct{Value: value}
}

// Nested package equivalent (nested functions/types)
func nestedFunction() string {
	return "Nested"
}

// Function that uses "parent" struct (equivalent to super import)
func useParentStruct() *HelperStruct {
	return NewHelperStruct(100)
}

// === CONFLICTING NAMES TEST ===
// Another set of helpers that also has a helperFunction (to test resolution)
func otherHelperFunction() int {
	return 42
}

// This function should call its own helper (local scope resolution)
func callOwnHelper() int {
	return otherHelperFunction() // Calls otherHelperFunction in same scope
}

// === DOT IMPORTS (equivalent to use statements that bring names into scope) ===
// In Go, we can simulate this by creating local aliases or using short variable names
var (
	// Local aliases (equivalent to bringing names into current scope)
	localHelperFunc = helperFunction
	LocalHelperStruct = HelperStruct
)

// === BLANK IMPORTS (side-effect only imports) ===
import (
	_ "net/http/pprof" // Side-effect import for profiling
)

// === MAIN FUNCTION WITH IMPORT USAGE ===
func main() {
	fmt.Println("=== Testing Import Resolution ===\n")

	// TEST 1: Standard library imports
	fmt.Println("1. Standard library imports:")
	
	// sync.Mutex from standard library
	var mu sync.Mutex
	mu.Lock()
	mu.Unlock()
	fmt.Println("   sync.Mutex created and used ✓")
	
	// time.Duration from standard library
	duration := time.Second
	fmt.Printf("   time.Duration used: %v ✓\n", duration)

	// TEST 2: Aliased imports
	fmt.Println("\n2. Aliased imports:")
	
	// Using aliased context as normal context
	ctx := context.Background()
	fmt.Printf("   context.Background() via import: %T ✓\n", ctx)
	
	// Using set alias (list as set-like structure)
	list := set.New()
	list.PushBack(1)
	fmt.Println("   list used as 'set' alias ✓")
	
	// Using lock alias for sync
	var lockMu lock.Mutex
	lockMu.Lock()
	lockMu.Unlock()
	fmt.Println("   sync.Mutex used as 'lock' alias ✓")

	// TEST 3: Local function resolution
	fmt.Println("\n3. Local function resolution:")
	result := helperFunction()
	fmt.Printf("   helperFunction() returns: '%s' (expected: 'Helper') ✓\n", result)
	
	helper := NewHelperStruct(10)
	fmt.Printf("   HelperStruct created with value: %d ✓\n", helper.Value)

	// TEST 4: Nested function calls
	fmt.Println("\n4. Nested function calls:")
	nestedResult := nestedFunction()
	fmt.Printf("   nestedFunction() returns: '%s' (expected: 'Nested') ✓\n", nestedResult)
	
	parentStruct := useParentStruct()
	fmt.Printf("   useParentStruct() returns struct with value: %d ✓\n", parentStruct.Value)

	// TEST 5: Local scope resolution vs conflicting names
	fmt.Println("\n5. Scope resolution with conflicting names:")
	// This should call helperFunction (not otherHelperFunction)
	localResult := helperFunction()
	fmt.Printf("   helperFunction() returns: '%s' (expected: 'Helper') ✓\n", localResult)
	
	// This should call otherHelperFunction through callOwnHelper
	otherResult := callOwnHelper()
	fmt.Printf("   callOwnHelper() returns: %d (expected: 42) ✓\n", otherResult)

	// TEST 6: Local aliases (equivalent to dot imports)
	fmt.Println("\n6. Local aliases (dot import equivalent):")
	aliasResult := localHelperFunc()
	fmt.Printf("   localHelperFunc() returns: '%s' (expected: 'Helper') ✓\n", aliasResult)
	
	aliasStruct := LocalHelperStruct{Value: 99}
	fmt.Printf("   LocalHelperStruct created with value: %d ✓\n", aliasStruct.Value)

	fmt.Println("\n=== All import tests completed ===")
}

// === TEST FUNCTIONS (equivalent to #[cfg(test)] mod tests) ===

func TestStandardImports() error {
	// Test that standard library types work
	var mu sync.Mutex
	mu.Lock()
	defer mu.Unlock()
	
	ctx := context.Background()
	if ctx == nil {
		return fmt.Errorf("context should not be nil")
	}
	
	return nil
}

func TestAliasedImports() error {
	// Test that aliases work correctly
	list := set.New() // Using aliased import
	if list == nil {
		return fmt.Errorf("aliased list should not be nil")
	}
	
	var lockMu lock.Mutex // Using aliased sync
	lockMu.Lock()
	defer lockMu.Unlock()
	
	return nil
}

func TestLocalResolution() error {
	// Test local function resolution
	result := helperFunction()
	if result != "Helper" {
		return fmt.Errorf("expected 'Helper', got '%s'", result)
	}
	
	// Test struct creation
	helper := NewHelperStruct(20)
	if helper.Value != 20 {
		return fmt.Errorf("expected value 20, got %d", helper.Value)
	}
	
	// Test nested function
	nested := nestedFunction()
	if nested != "Nested" {
		return fmt.Errorf("expected 'Nested', got '%s'", nested)
	}
	
	return nil
}

func TestScopeResolution() error {
	// Test that we call the right function in the right scope
	result := callOwnHelper()
	if result != 42 {
		return fmt.Errorf("expected 42, got %d", result)
	}
	
	return nil
}

// Nested test functions (equivalent to nested test modules)
func TestNestedResolution() error {
	// Test accessing "grandparent" scope
	result := helperFunction()
	if result != "Helper" {
		return fmt.Errorf("nested test: expected 'Helper', got '%s'", result)
	}
	
	return nil
}

// Helper function to run all tests
func RunAllTests() {
	tests := []struct {
		name string
		test func() error
	}{
		{"Standard Imports", TestStandardImports},
		{"Aliased Imports", TestAliasedImports},
		{"Local Resolution", TestLocalResolution},
		{"Scope Resolution", TestScopeResolution},
		{"Nested Resolution", TestNestedResolution},
	}
	
	fmt.Println("\n=== Running Tests ===")
	allPassed := true
	
	for _, test := range tests {
		if err := test.test(); err != nil {
			fmt.Printf("❌ %s: %v\n", test.name, err)
			allPassed = false
		} else {
			fmt.Printf("✅ %s: passed\n", test.name)
		}
	}
	
	if allPassed {
		fmt.Println("\n🎉 All tests passed!")
	} else {
		fmt.Println("\n❌ Some tests failed!")
	}
}

// === RESOLUTION EXPECTATIONS SUMMARY ===
//
// When indexed and analyzed, we expect:
//
// 1. STANDARD LIBRARY:
//    - fmt resolves to standard library fmt package
//    - sync resolves to standard library sync package
//    - time resolves to standard library time package
//    - context resolves to standard library context package
//
// 2. ALIASES:
//    - set resolves to container/list (with alias tracked)
//    - lock resolves to sync (with alias tracked)
//
// 3. LOCAL FUNCTIONS:
//    - helperFunction resolves to local helperFunction in main package
//    - HelperStruct resolves to local HelperStruct type
//    - NewHelperStruct resolves to local constructor function
//
// 4. NESTED CALLS:
//    - nestedFunction resolves to local nestedFunction
//    - useParentStruct resolves to local useParentStruct
//
// 5. SCOPE RESOLUTION:
//    - helperFunction() in main calls local helperFunction
//    - otherHelperFunction() in callOwnHelper calls otherHelperFunction
//    - Local vs package scope should be correctly distinguished
//
// 6. LOCAL ALIASES:
//    - localHelperFunc resolves to helperFunction via variable alias
//    - LocalHelperStruct resolves to HelperStruct via type alias
//
// 7. BLANK IMPORTS:
//    - _ "net/http/pprof" should be tracked as side-effect import
//
// 8. DISAMBIGUATION:
//    - helperFunction() calls local helperFunction (direct reference)
//    - callOwnHelper() -> otherHelperFunction() (local scope)
//    - Multiple functions with same name in different scopes resolved correctly

// Run tests when this file is executed
func init() {
	// Automatically run tests when package is loaded
	go func() {
		time.Sleep(100 * time.Millisecond) // Let main run first
		RunAllTests()
	}()
}